gnucash/gnome/gnc-split-reg.c

```diff
@@ -1054,33 +1055,50 @@ gsr_default_associate_handler_file (GNCSplitReg *gsr, Transaction *trans, gboole
 
     gtk_file_chooser_set_local_only (GTK_FILE_CHOOSER(dialog), FALSE);
 
-    if ((path_head != NULL) && (g_strcmp0 (path_head, "") != 0)) // not default entry
-        valid_path_head = TRUE;
+    if (path_head && g_strcmp0 (path_head, "") != 0) // not default entry
+        path_head_set = TRUE;
 
     if (have_uri)
     {
-        gchar *new_uri;
-        gchar *uri_label;
-        gchar *filename;
-
+        gchar *file_uri = NULL;
         const gchar *uri = xaccTransGetAssociation (trans);
+        gchar *scheme = gnc_uri_get_scheme (uri);
 
-        if (valid_path_head && g_str_has_prefix (uri,"file:/") && !g_str_has_prefix (uri,"file://"))
+        if (!scheme) // relative path
         {
-            const gchar *part = uri + strlen ("file:");
-            new_uri = g_strconcat (path_head, part, NULL);
+            gchar *file_path = NULL;
+            if (path_head_set) // not default entry
+                file_path = gnc_file_path_absolute (gnc_uri_get_path (path_head), uri);
+            else
+                file_path = gnc_file_path_absolute (NULL, uri);
+
+            file_uri = gnc_uri_create_uri ("file", NULL, 0, NULL, NULL, file_path);
+            g_free (file_path);
         }
-        else
-            new_uri = g_strdup (uri);
 
-        filename = g_uri_unescape_string (new_uri, NULL);
-        uri_label = g_strconcat (_("Existing Association is "), filename, NULL);
-        gtk_file_chooser_set_extra_widget (GTK_FILE_CHOOSER(dialog), gtk_label_new (uri_label));
-        gtk_file_chooser_set_uri (GTK_FILE_CHOOSER(dialog), new_uri);
+        if (g_strcmp0 (scheme, "file") == 0) // absolute path
```

Comment:
I know the implicit assumption is that file associations can only be set via a GtkFileChooserDialog, hence file is the only scheme to expect. However in the future other parts of gnucash may also set file associations (like python bindings) and they may be using another scheme as supported by gnc_uri_utils (sqlite or xml come to mind). So to be future safe it's probably better to use the gnc_uri_utils provided function to test whether the given scheme is file compatible.